_posts/2024-09-25-Testing-Scientific-Software.md

```diff
@@ -0,0 +1,374 @@
+---
+title: "Testing Scientific Software: A Practical Guide for Developers"
+author: dave-bunten
+tags:
+  - software-testing
+  - scientific-software
+  - python
+  - reproducibility
+---
+
+# Testing Scientific Software: A Practical Guide for Developers
+
+{% include blog-post-intro.html %}
+
+## Introduction
+
+{% include figure.html image="images/magnifying-glass-and-bugs.png" width="40%" caption="Software testing helps us find bugs in our code which otherwise may go unseen." %}
+
+<!-- excerpt start -->
+**Scientific software plays a crucial role in research.**
+When your software is used to analyze data, simulate models, or derive scientific conclusions, ensuring its correctness becomes critical.
+**A small bug can have a significant impact on your results**, potentially invalidating months of work, or worse, causing the retraction of published research (for example: [see here](https://www.science.org/doi/10.1126/science.314.5807.1856)).
+Fortunately, software testing can help minimize such risks, giving you more confidence in your code and a greater chance to catch issues early.
+
+In this guide, we’ll walk through key types of software tests, practical advice for using popular testing tools like `pytest` and `doctest`, and how you can incorporate these into your scientific development workflow.
+<!-- excerpt end -->
+
+## Why testing matters in science
+
+Imagine you’ve written a simulation that generates data based on a complex scientific model.
+It works well under some conditions, but during peer review, a colleague finds a subtle bug.
+This bug doesn't affect small data sets but produces significant errors on larger simulations.
+The consequences?
+You have to revise your paper, and time is lost fixing code.
+There's also the possibility of bugs taking a long time to find (if ever) potentially leading to erroneous research.
+
+Testing your software upfront ensures that potential errors are caught early and your scientific conclusions remain valid and robust.
+
```

Comment:
I'd like to see a reference, perhaps here, to the fact that people do typically "test" their software, just not automatically. E.g., running your software and validating the output manually is a kind of test, and one that anyone who writes software already does. The article IMHO is about automated testing, which helps ensure that the tests are applied frequently and consistently. I'd make the case that automating testing ends up saving researchers not only time and expense from missed bugs, but also the toil and worry of having to shoulder the responsibility of testing manually themselves.

Reply:
Thanks good point! I feel this is about both the act of testing and automating that testing. The trouble with "testing without automation / code" is that it often is undeclared or declared in a bespoke manner. This is I think related to making unfounded claims; how can an audience know a software creates a reproducible outcome without tests (how could we prove a hypothesis without conducting an experiment someone else could recreate)? I'll add something about this in this part of the article.

Reply:
I totally agree. I think the point of my comment was to suggest that what they're currently doing with ad-hoc testing isn't so different from automated testing, except that it's all the things you said: manual, bespoke, and not part of the published code and thus not able to provide proof of their claims that the software works. I was thinking it might not be so daunting to implement tests if they think of it as automating what they're likely already doing manually.

Reply:
That makes sense, probably there is a bit of fear here we need to address! Rearranged / worded things here based on this.